server/src/main/java/org/elasticsearch/index/mapper/SyntheticSourceDocumentParserListener.java

```diff
@@ -0,0 +1,192 @@
+                            // does not feel great though
+                            valuesToStore.add(
+                                new IgnoredSourceFieldMapper.NameValue(fullPath, parentOffset, XContentDataHelper.encodeXContentBuilder(data), document)
+                            );
+                            return returnState;
+                        }
+                    }
+                }
+                case Token.FieldName fieldName -> data.field(fieldName.name());
+                case Token.StringValue stringValue -> data.value(stringValue.value());
+                case Token.BooleanValue booleanValue -> data.value(booleanValue.value());
+                case Token.IntValue intValue -> data.value(intValue.value());
+            }
+
+            return this;
+        }
+    }
+
+    class Watching implements State {
+        private String fullPath;
+        private Stack<ObjectMapper> parents;
+        private Stack<String> objectArrays;
+        private Mapper currentMapper;
+
+        Watching(String fullPath, Stack<ObjectMapper> parents, Mapper currentMapper) {
+            this.fullPath = fullPath;
+            this.parents = parents;
+            this.objectArrays = new Stack<>();
+            this.currentMapper = currentMapper;
+        }
+
+        public State consume(Token token) throws IOException {
+            switch (token) {
+                case Token.StartObject startObject -> {
+                    if (currentMapper instanceof ObjectMapper om && om.isEnabled() == false) {
+                        var storingState = new Storing(this, startObject, fullPath, parents.peek());
+                        // TODO should we some cleaner "reset()" method on Watching or something?
+                        currentMapper = null;
+                        fullPath = null;
+                        return storingState;
+                    }
+                    if (currentMapper instanceof ObjectMapper om) {
+                        parents.push(om);
+                        currentMapper = null;
+                    }
+                }
+                case Token.EndObject endObject -> {
+                    parents.pop();
+                }
+                case Token.StartArray startArray -> {
+                    if (currentMapper instanceof ObjectMapper om) {
+                        objectArrays.push(fullPath);
+                    }
+                }
+                case Token.EndArray endArray -> {
+                    if (currentMapper instanceof ObjectMapper om) {
+                        objectArrays.push(fullPath);
+                    }
+                }
+                case Token.FieldName fieldName -> {
+                    ObjectMapper parentMapper = parents.peek();
+                    fullPath = parentMapper.isRoot() ? fieldName.name() : parentMapper.fullPath() + "." + fieldName.name();
+                    currentMapper = parentMapper.getMapper(fieldName.name());
```

Comment:
This doesn't work for dynamic mappers. Do we have an issue with that? 

I guess we don't use ignored source for dynamic objects, only their parents.

Reply:
I think it should be okay? If this returns `null` and parent has something like `dynamic: false` we go into storing mode?